src/inmanta/deploy/tasks.py

```diff
@@ -45,13 +46,17 @@ class Task(abc.ABC):
     """
 
     resource: ResourceIdStr
-
     id: resources.Id = dataclasses.field(init=False, compare=False, hash=False)
+    reason: typing.Optional[str] = dataclasses.field(compare=False, hash=False, default=None)
 
     def __post_init__(self) -> None:
         # use object.__setattr__ because this is a frozen dataclass, see dataclasses docs
         object.__setattr__(self, "id", resources.Id.parse_id(self.resource))
 
+    def set_reason(self, reason: str) -> None:
+        # use object.__setattr__ because this is a frozen dataclass, see dataclasses docs
+        object.__setattr__(self, "reason", reason)
```

Comment:
This is something we should not do. It will change the hash of the object, which is rather bad, since this object is used as the key of the dictionary.

```
>>> import abc
>>> from dataclasses import dataclass
>>> @dataclass(frozen=True, kw_only=True)
... class Task(abc.ABC):
...     x: str
... 
>>> t = Task(x="test")
>>> dct = {}
>>> dct[t] = "aaa"
>>> 
>>> object.__setattr__(t, "x", "bbb")
>>> dct[t]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: Task(x='bbb'')
```

Reply:
There is a subtilty that was not highlighted enough: It's not taken into consideration in the hash of the object
```
>>> import abc
>>> import dataclasses
>>> from dataclasses import dataclass
>>> @dataclass(frozen=True, kw_only=True)
... class Task(abc.ABC):
...     x: str
...     y: typing.Optional[str] = dataclasses.field(compare=False, hash=False, default=None)
... 
>>> t = Task(x="test")
>>> dct = {}
>>> dct[t] = "aaa"
>>> object.__setattr__(t, "y", "bbb")
>>> dct[t]
'aaa'
>>> 
```